Support for json:",omitzero" tag

nsd20463 opened the pull request:
This also fixes the 
  X uintptr  json:",omitempty" 
omitempty applied to a zero-value uintptr is treated as empty by the stdlib json.Marshal, but wasn't by easyjson.

nsd20463 committed:
add support for the json:",omitzero" tag

omitzero was added to the go stdlib encoder/json in go 1.24

nsd20463 committed:
add case which omitempty missed: uintptr is encodable in json

and does accept omitempty in the go stdlib encoder/json.

neal commented on tests/data.go:
```diff
@@ -358,6 +358,39 @@ var omitEmptyString = "{" +
+
+	// NOTE: omitzero DOES have effect on non-pointer struct fields.
+	SubZ, SubNZ   SubStruct  `json:",omitzero"`
+	SubPZ, SubPNZ *SubStruct `json:",omitzero"`
```
may be worth also testing pointer values set with zero-value struct/string

nsd20463 replied:
sure, why not. done.

nsd20463 committed:
add omitzero test cases for pointer-to-"" and pointer-to-{}

(both should marshal to something in order to match stdlib json, and
they do)

nsd20463 committed:
fix and add tests for the case when omitzero is combined with omitempty

Without a special case the generated code can look like

   if v != nil && v != nil {
or
   if v != "" && v != "" {

and that gets flagged by go vet (go 1.25.5)

nsd20463 commented:
While adding more test cases I realized that combining omitempty and omitzero together generated code which didn't pass `go vet`, so the most recent commit fixes that by treating it as a special case in the encoder generator.

neal reviewed (approved):
LGTM

YesYouKenSpace commented:
Any way we can merge this in?
